test/unit/db-adapters/Mongoose/MongooseAdapter.js

```diff
@@ -69,68 +74,190 @@ describe("Mongoose Adapter", () => {
-          expect(res[0]).to.equal("find");
-          expect(res[1]._id.$in).to.be.an.Array;
-          expect(res[1]._id.$in[0]).to.equal("552c5e1c604d41e5836bb174");
-          expect(res[1]._id.$in[1]).to.equal("552c5e1c604d41e5836bb175");
+        it("should throw if any numeric ids are invalid", function(done) {
+          MongooseAdapter.getIdQueryType(["abc", "123"], NumericId).then(res => {
+            expect(false).to.be.ok;
+          }, err => {
+            expect(err).to.be.instanceof(APIError);
+            done();
+          }).catch(done);
+        });
+
+        it("should produce query on valid ObjectId input", function(done) {
+          MongooseAdapter.getIdQueryType(["552c5e1c604d41e5836bb174", "552c5e1c604d41e5836bb175"], School).then(([ mode, idQuery ]) => {
+            expect(mode).to.equal("find");
+            expect(idQuery._id.$in).to.be.an.Array;
+            expect(idQuery._id.$in[0]).to.equal("552c5e1c604d41e5836bb174");
+            expect(idQuery._id.$in[1]).to.equal("552c5e1c604d41e5836bb175");
+            done();
+          }).catch(done);
+        });
+
+        it("should produce query on valid numeric ID input", function(done) {
+          MongooseAdapter.getIdQueryType(["0", "1"], NumericId).then(([ mode, idQuery ]) => {
+            expect(mode).to.equal("find");
+            expect(idQuery._id.$in).to.be.an.Array;
+            expect(idQuery._id.$in[0]).to.equal("0");
+            expect(idQuery._id.$in[1]).to.equal("1");
+            done();
+          }).catch(done);
+        });
+
+        it("should produce query on valid string ID input", function(done) {
+          MongooseAdapter.getIdQueryType(["a", "null"], StringId).then(([ mode, idQuery ]) => {
+            expect(mode).to.equal("find");
+            expect(idQuery._id.$in).to.be.an.Array;
+            expect(idQuery._id.$in[0]).to.equal("a");
+            expect(idQuery._id.$in[1]).to.equal("null");
+            done();
+          }).catch(done);
         });
       });
     });
 
-    describe("idIsValid", () => {
-      it("should reject all == null input", () => {
-        expect(MongooseAdapter.idIsValid()).to.not.be.ok;
-        expect(MongooseAdapter.idIsValid(null)).to.not.be.ok;
-        expect(MongooseAdapter.idIsValid(undefined)).to.not.be.ok;
+    describe("validateId", () => {
+      it("should refuse all == null input", function(done) {
+        const tests = [
+          MongooseAdapter.validateId(null, School),
+          MongooseAdapter.validateId(null, NumericId),
+          MongooseAdapter.validateId(null, StringId),
+          MongooseAdapter.validateId(undefined, School),
+          MongooseAdapter.validateId(undefined, NumericId),
+          MongooseAdapter.validateId(undefined, StringId)
+        ];
+
+        Q.allSettled(tests).then((res) => {
+          res.forEach(result => expect(result.state).to.equal("rejected"));
```

Comment:
Well yes, it's just that's a very common pattern, so common that we have third party libraries that do that and lots of other useful things. I'm not sure what value you get out of reimplementing.